### PR TITLE
add package-info to cluster.routing

### PR DESCRIPTION
**Pull request by mhl-b**
Add package-info to cluster routing. Does not include `allocation` package information, but it might be better to have routing and allocation description in one place.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * shards in cluster {@link org.elasticsearch.cluster.routing.RoutingTable} and
+ * {@link org.elasticsearch.cluster.routing.RoutingNodes}. RoutingTable provides
+ * view from index to shard to node, and RoutingNodes provides view from node to
+ * shard. Shard allocation is a process of assigning and moving shards between
+ * nodes. For more details about allocation see {@link
+ * org.elasticsearch.cluster.routing.allocation}.</p>
+ *
+ * <b>Routing Tables</b>
+ *
+ * <p>Routing tables provide a view on global cluster state from index
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:21]:
```suggestion
 * <p>The routing table provide a view on global cluster state from index
```

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * perspective. It's a mapping from indices to shards and shards to nodes, where
+ * relationship between shard and node might not exists if shard allocation is
+ * not possible. Routing table represents desired shards layout and current
+ * state of shards.</p>
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:25]: I don't understand this last sentence. The previous sentence states that a shard-to-node mapping may not exist (if the shard has not been allocated yet). And then we're talking about desired shard allocation being in the routing table?

Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:25]: "desired" word might be confusing, its desired from allocator perspective. If allocator cannot find right place for shard it will be unassigned.

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * org.elasticsearch.cluster.routing.IndexShardRoutingTable}. Each shard can
+ * have one or more instances - primary and replicas. IndexShardRoutingTable
+ * contains information about all shard instances for specific shard id - {@link
+ * org.elasticsearch.cluster.routing.ShardRouting}.</p>
+ *
+ * <b>Routing Nodes</b>
+ *
+ * <p>{@link org.elasticsearch.cluster.routing.RoutingNode} provides view to
+ * shard routing from node perspective. Each RoutingNode is a {@link
+ * org.elasticsearch.cluster.node.DiscoveryNode}, and contains information about
```
[the remark this answers is not in the document]
Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:42]: thanks @henningandersen!

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * <p>As class hierarchy goes - at the top level there is a {@link
+ * org.elasticsearch.cluster.routing.RoutingTable}. RoutingTable is a part of
+ * {@link org.elasticsearch.cluster.ClusterState}. It contains mapping between
+ * indices and {@link org.elasticsearch.cluster.routing.IndexRoutingTable}.
+ * {@link org.elasticsearch.cluster.routing.IndexRoutingTable} contains routing
+ * information about all shards in that index - {@link
+ * org.elasticsearch.cluster.routing.IndexShardRoutingTable}. Each shard can
+ * have one or more instances - primary and replicas. IndexShardRoutingTable
+ * contains information about all shard instances for specific shard id - {@link
+ * org.elasticsearch.cluster.routing.ShardRouting}.</p>
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:36]: ShardRouting is kind of thrown in here. Perhaps add a little more elaboration?

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * <b>Routing Nodes</b>
+ *
+ * <p>{@link org.elasticsearch.cluster.routing.RoutingNode} provides view to
+ * shard routing from node perspective. There is a RoutingNode for every {@link
+ * org.elasticsearch.cluster.node.DiscoveryNode}, and it contains information about
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:42]: I don't see a RoutingNode object in DiscoveryNode? Is there something more to this relationship that could be said here?

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java (hunk 2 above)
Comment (DiannaHohensee) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:25]: So where is the desired allocation information saved? Presumably not the same place as the actual mappings? Or is it all in one data structure and there is some kind of flag saying that the shard in question is not allocated?

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:25]: I would suggest to rephrase it to `Routing table represents current state of shards`

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * <b>Reroute</b>
+ *
+ * <p>Reroute is a process of routing change in the cluster. It will start
+ * allocation process, which might assign or move shards around. When cluster
+ * has an update that impacts routing (for example: node join cluster, index
+ * created, snapshot restored, ...) the reroute process must be triggered. There
+ * are 2 ways to trigger reroute - batched(recommended) with {@link
+ * org.elasticsearch.cluster.routing.BatchedRerouteService} and non-batch {@link
+ * org.elasticsearch.cluster.routing.allocation.AllocationService}.</p>
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:56]: I would say it is a little more complex than that. Batched can combine multiple requests into one (used when starting initializing shards), but unbatched one allows to  mix other cluster state changes that might be required to create or delete index.

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java (hunk 2 above)
Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:25]: @DiannaHohensee, my understanding that allocator will create assign shards to nodes and produce new routing table. When node receive cluster update it will see which shards are allocated there and will start initialization process - shard state machine. So routing table is a desired layout in terms of shards assignment and current state of a shard at the same time.
And I agree with @idegtiarenko, probably better to keep it simple here. And move allocation part into "future" cluster.routing.allocation package-info

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * <p>Cluster routing package provides routing information and manages shard
+ * allocations in cluster. The routing part contains two different views on
+ * shards in cluster {@link org.elasticsearch.cluster.routing.RoutingTable} and
+ * {@link org.elasticsearch.cluster.routing.RoutingNodes}. RoutingTable provides
+ * view from index to shard to node, and RoutingNodes provides view from node to
+ * shard. Shard allocation is a process of assigning and moving shards between
+ * nodes. For more details about allocation see {@link
+ * org.elasticsearch.cluster.routing.allocation}.</p>
+ *
+ * <b>Routing Table</b>
```
[the remark this answers is not in the document]
Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:19]: RoutingTable, RoutingNodes and reroute, they have own documentation. It wasnt my intention to replace or duplicate what existing docs say, but rather point out how they interact which each other. It might be good to have some of these notes duplicated at class/method level. May be next PR

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java (hunk 5 above)
Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java]: A RoutingNode has association with DiscoveryNode, not vice versa.

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java (hunk 4 above)
Comment (DiannaHohensee) [reply server/src/main/java/org/elasticsearch/cluster/routing/package-info.java]: Thanks!

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ *         │   │       └── UNASSIGNED
+ *         │   └── shard2-IndexShardRoutingTable ...
+ *         └── index2-IndexRoutingTable ...
+ * }
+ * </pre>
+ *
+ *
+ * <b>Routing Nodes</b>
+ *
+ * <p>{@link org.elasticsearch.cluster.routing.RoutingNode} provides a view to
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:62]:
```suggestion
 * <p>{@link org.elasticsearch.cluster.routing.RoutingNode} provides a view
```

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ * shards in {@link org.elasticsearch.cluster.routing.RoutingTable} and
+ * {@link org.elasticsearch.cluster.routing.RoutingNodes}. RoutingTable provides
+ * a view from index to shard to node. RoutingNodes provides view from node
+ * shard. Shard allocation is a process of assigning and moving shards between
+ * nodes. For more details about allocation see {@link
+ * org.elasticsearch.cluster.routing.allocation}.</p>
+ *
+ * <b>Routing Table</b>
+ *
+ * <p>The routing table provide a view of global cluster state from index
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:21]:
```suggestion
 * <p>The routing table provide a view of global cluster state from an index
```

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+ * perspective. It's a mapping from indices to shards and shards to nodes, where
+ * relationship between shard and node may not exists if a shard allocation is
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:23]:
```suggestion
 * the relationship between shard and node may not exist if a shard allocation is
```

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+ * <p>{@link org.elasticsearch.cluster.routing.RoutingTable} is the access
+ * point to the routing information. RoutingTable is a part of the
+ * {@link org.elasticsearch.cluster.ClusterState}. It maps index names to
+ * {@link org.elasticsearch.cluster.routing.IndexRoutingTable}.
+ * {@link org.elasticsearch.cluster.routing.IndexRoutingTable}, in turn,
+ * holds a list of shards in that index,
+ * {@link org.elasticsearch.cluster.routing.IndexShardRoutingTable}. Each
+ * shard has one or more instances: a primary and replicas. An
+ * IndexShardRoutingTable contains information about all shard instances for
+ * specific shard, {@link org.elasticsearch.cluster.routing.ShardRouting}.
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:35]:
```suggestion
 * a specific shard, {@link org.elasticsearch.cluster.routing.ShardRouting}.
```

File: server/src/main/java/org/elasticsearch/cluster/routing/package-info.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+ * The ShardRouting is a state of a shard instance in a cluster. There are
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/cluster/routing/package-info.java:36]:
```suggestion
 * The ShardRouting is the state of a shard instance in a cluster. There are
```

**Review by DiannaHohensee — APPROVED** [review]
Added some additional grammatical suggestions.

Looks great, thanks for writing it up!
